File: Kotlin/sort/quick_sort.kt
fun quick_sort(A: Array<Int>, p: Int, r: Int) {
    if (p < r) {
        var q: Int = partition(A, p, r)
        quick_sort(A, p, q - 1)
        quick_sort(A, q + 1, r)

    }
}

fun partition(A: Array<Int>, p: Int, r: Int): Int {
    var x = A[r]
    var i = p - 1
    for (j in p until r) {
        if (A[j] <= x) {
            i++
            exchange(A, i, j)
        }
    }
    exchange(A, i + 1, r)
    return i + 1
}

fun exchange(A: Array<Int>, i: Int, j: Int) {
    var temp = A[i]
    A[i] = A[j]
    A[j] = temp
}

fun main(arg: Array<String>) {
    print("Enter no. of elements :")
    var n = readLine()!!.toInt()

    println("Enter elements : ")
    var A = Array(n, { 0 })
    for (i in 0 until n)
        A[i] = readLine()!!.toInt()

    quick_sort(A, 0, A.size - 1)

    println("Sorted array is : ")
    for (i in 0 until n)
        print("${A[i]}  ")
}

/*
Output: 

Enter no. of elements: 5
Enter elements:
10
34
22
56
74
Sorted array is:
10  22  34  56  74

*/

File: Dart/sort/Quick_Sort.dart
/* 
Quick Sort in Dart

Quicksort algorithm is based on the divide and conquer approach where
  1. An array is divided into subarrays by selecting a pivot element.
  2. The left and right subarrays are also divided using the same approach. 
    This process continues until each subarray contains a single element.
  3. At this point, elements are already sorted. Finally, elements are 
    combined to form a sorted array.
*/

import 'dart:io';

// function to find the partition position
int partition(var array, int low, int high) {
  // choosing rightmost element as pivot
  int pivot = array[high];
  // pointer for greater element
  int i = low - 1;

  for (int j = low; j < high; j++) {
    // compare each element with pivot
    if (array[j] <= pivot) {
      i = i + 1;
      // if element smaller than pivot is found swap it i
      int temp = array[j];
      array[j] = array[i];
      array[i] = temp;
    }
  }
  // swapping pivot with greater element
  int temp = array[i + 1];
  array[i + 1] = array[high];
  array[high] = temp;

  // return the position from where partition is done
  return i + 1;
}

// function to perform quicksort
void quickSort(array, low, high) {
  if (low < high) {
    // find pivot element such that
    // element smaller than pivot are on left and greater than are at right
    int pi = partition(array, low, high);
    quickSort(array, low, pi - 1);
    quickSort(array, pi + 1, high);
  }
}

// main function
void main() {
  var data = [];

  // taking input from user and storing it in data array
  var n, ele;
  print('Enter the number of Elements: ');
  n = stdin.readLineSync();
  n = int.parse(n);

  for (int i = 1; i <= n; i++) {
    print('Enter Element $i: ');
    ele = stdin.readLineSync();
    ele = int.parse(ele);
    data.add(ele);
  }

  // Implementing quick sort
  quickSort(data, 0, n - 1);

  // displaying sorted array
  print(data);
}

/*

Time Complexities:
  1. Worst Case Complexity : O(n^2)
  2. Best Case Complexity : O(n*log n)
  3. Average Case Complexity : O(n*log n)

Space Complexity : O(log n)

INPUT :-
Enter the number of Elements: 
5
Enter Element 1: 
20
Enter Element 2: 
10
Enter Element 3: 
44
Enter Element 4:
33
Enter Element 5:
99

Output :-
[10, 20, 33, 44, 99]
*/
